core/aws-core/src/main/java/software/amazon/awssdk/awscore/internal/authcontext/AwsCredentialsAuthorizationStrategy.java

```diff
@@ -84,22 +93,27 @@ public void addCredentialsToExecutionAttributes(ExecutionAttributes executionAtt
      *
      * @return The credentials provider that will be used by the SDK to resolve credentials
      */
-    private static AwsCredentialsProvider resolveCredentialsProvider(SdkRequest originalRequest,
-                                                                    AwsCredentialsProvider defaultProvider) {
+    private static IdentityProvider<? extends AwsCredentialsIdentity> resolveCredentialsProvider(
+            SdkRequest originalRequest,
+            IdentityProvider<? extends AwsCredentialsIdentity> defaultProvider) {
         return originalRequest.overrideConfiguration()
                               .filter(c -> c instanceof AwsRequestOverrideConfiguration)
                               .map(c -> (AwsRequestOverrideConfiguration) c)
-                              .flatMap(AwsRequestOverrideConfiguration::credentialsProvider)
+                              .flatMap(AwsRequestOverrideConfiguration::credentialsIdentityProvider)
                               .orElse(defaultProvider);
     }
 
-    private static AwsCredentials resolveCredentials(AwsCredentialsProvider credentialsProvider,
-                                                     MetricCollector metricCollector) {
+    private static AwsCredentialsIdentity resolveCredentials(
+            IdentityProvider<? extends AwsCredentialsIdentity> credentialsProvider,
+            MetricCollector metricCollector) {
         Validate.notNull(credentialsProvider, "No credentials provider exists to resolve credentials from.");
 
-        Pair<AwsCredentials, Duration> measured = MetricUtils.measureDuration(credentialsProvider::resolveCredentials);
+        // TODO: Exception handling for join()?
+        Pair<? extends AwsCredentialsIdentity, Duration> measured =
+            MetricUtils.measureDuration(() -> credentialsProvider.resolveIdentity().join());
```

Comment:
Can we avoid joining at all for async? Or is that too much work (a TODO)? 

If we need to join for now, can we use CompletableFutureUtils.joinInterruptibly and then unwrap CompletionExceptions that have a RuntimeException or Error cause?

e.g.

```java
CompletableFutureUtils.joinLikeSync(credentialsProvider::resolveIdentity);

/**
 * Joins (interruptibly) on the future, and re-throws any RuntimeExceptions or Errors just like the async task would have thrown if it was executed synchronously.
 */
<T> T joinLikeSync(CompletableFuture<T> future) {
    try {
        return CompletableFutureUtils.joinInterruptibly(future);
    } catch (CompletionException e) {
        Throwable cause = e.getCause();
        if (cause instanceof RuntimeException) {
            cause.addSuppressed(new CompletionException("Task failed.")); // Make sure we don't lose the context of where the join is in the stack...
            throw (RuntimeException) cause;
        }
    }
}
```

Reply:
Discussed offline, will leave all `join()`s in this PR for now, but will revisit how to convert these to async where possible before we are decide on releasing the Identity changes.